spec/abstract/KHeap_A.thy

```diff
@@ -11,12 +11,55 @@ Functions to access kernel memory.
 chapter \<open>Accessing the Kernel Heap\<close>
 
 theory KHeap_A
-imports Exceptions_A
+imports Exceptions_A Lib.SplitRule
 begin
 
+section "Projection from the Kernel Heap"
+
 text \<open>This theory gives auxiliary getter and setter methods
 for kernel objects.\<close>
 
+section "Projection from the Kernel Heap"
+
+definition tcb_of :: "kernel_object \<rightharpoonup> tcb" where
+  "tcb_of ko \<equiv> case ko of TCB tcb \<Rightarrow> Some tcb | _ \<Rightarrow> None"
+
+lemmas tcb_of_simps [simp] = tcb_of_def [split_simps kernel_object.split]
+
+lemma tcb_of_Some[simp]:
+  "tcb_of ko = Some tcb \<longleftrightarrow> ko = TCB tcb"
+  by (cases ko; simp)
+
+lemma tcb_of_None:
+  "tcb_of ko = None \<longleftrightarrow> (\<forall>tcb. ko \<noteq> TCB tcb)"
+  by (cases ko; simp)
+
+definition tcbs_of_kh :: "('obj_ref \<rightharpoonup> kernel_object) \<Rightarrow> 'obj_ref \<rightharpoonup> tcb" where
+  "tcbs_of_kh kh \<equiv> kh |> tcb_of"
+
+abbreviation tcbs_of :: "'z state \<Rightarrow> obj_ref \<rightharpoonup> tcb" where
+  "tcbs_of s \<equiv> tcbs_of_kh (kheap s)"
```

Comment:
We don't usually do that additional wrapping in a definition for `kh`. It'll interfere somewhat with the projection stack, because the simplifier can't see the `|>` in the final term and can't apply the generic projection rules. Is there any specific reason for it or is it just what was there before? It looks a bit like the locales version of projections.

We could make it an abbreviation if we just want to have the name.

Reply:
This stuff was just copied, and I’m quite sure it was written by Matt, so maybe there are some difference in intention and approach here. I don’t think having the name would be helpful. Could you maybe write out exactly what you’d like? Or I can take a stab at it. 

I also wasn’t sure whether to commit to having all the projections in this file. I can move the others here if we’d like to do that

Reply:
Hm. That might turn into a greater rabbit hole than we like. If it's Matt's locales then we should move the locale setup as well, and then there is the question if we really want to keep that if we're doing the other projection style elsewhere. In terms of types and spec and expressivity they are the same, so for ASpec it's not really a problem, but I think I would only do the ones we need for now.

Reply:
Yes, thinking about it some more, it probably is a rabbit hole we want to avoid for now. I think if we were to get rid of anything here, then it's very likely to break some of the very sensitive simplification rules that are part of the locales, and it would be a monstrous task to fix that up. 

Should we just leave it here, maybe adding a comment before the definitions of `tcbs_of_kh` and `scs_of_kh` saying "Consider eliminating in favour of a direct definition below in terms of `kheap`"?

Reply:
Yes, that sounds good. Maybe also a hint that there is locale setup for this later, so that we don't forget that this exists.